### PR TITLE
Implement table `new` and `close`

### PR DESCRIPTION
**Pull request by jorisdral**

### CODE REVIEW

File: src/Database/LSMTree/Normal.hs [hunk 1]
```diff
@@ -167,8 +167,9 @@ import           Database.LSMTree.Internal.Normal
 -- such instances in use at once.
 type TableHandle :: (Type -> Type) -> Type -> Type -> Type -> Type
 data TableHandle m k v blob = forall h. Typeable h =>
-    TableHandle (Internal.TableHandle m h)
+    TableHandle {-# UNPACK #-} !(Internal.TableHandle m h)
```
Comment (mheinzel) [inline src/Database/LSMTree/Normal.hs:170]: The `UNPACK` pragma goes through newtypes here, so you are unpacking the `RWVar` fields into here?

I'm always a bit cautious about `UNPACK`. Without looking into the Core I wouldn't be sure it's an improvement, i.e. there is no function being called down the line that takes an internal handle or `RWVar` for which it needs to be reboxed.

File: src/Database/LSMTree/Internal.hs [hunk 2]
```diff
@@ -235,18 +255,36 @@ closeSession Session{sessionState} =
 --
 -- For more information, see 'Database.LSMTree.Normal.TableHandle'.
 --
--- NOTE: a table handle is __not__ thread-safe.
-data TableHandle m h = TableHandle {
-    tableSession     :: !(Session m h)
+-- Concurrent access to a table handle is mediated by a multiple-read,
+-- single-write lock. This allows concurrent read access, but only sequential
+-- write access. TODO: more fine-grained concurrent access than a top-level
+-- read-write lock.
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal.hs:261]: Do I understand this correctly: This comment is not talking about read and write access in the sense of lookups and inserts, but reading and writing the handle open/close state? Every operation except `close` would be considered read access here (but can of course still do modifications to the internal `Mvar`s)?

**Review by mheinzel — APPROVED** [review]

File: src/Database/LSMTree/Internal.hs (hunk 2 above)
Comment (jorisdral) [reply src/Database/LSMTree/Internal.hs:261]: The intended use would be to acquire a write lock for anything that updates any state in the handle, so not just `TableHandleState`. However, I kept the `MVar`s in there so we can implement more fine-grained concurrent access later. I can remove the MVars for now if that makes it clearer

File: src/Database/LSMTree/Normal.hs (hunk 1 above)
Comment (jorisdral) [reply src/Database/LSMTree/Normal.hs:170]: I've removed the `UNPACK` for now. We can add them back later if we think it's worth it

File: src/Database/LSMTree/Internal.hs (hunk 2 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal.hs:261]: That's very coarse-grained, but okay for a first version.

**Pull request merged**